### ISSUE TITLE
Undersanding Server Order

### ISSUE THREAD

**Issue by vbisbest** [MichaCo/DnsClient.NET#18]
Opened new issue, but more of a question.  I was hoping the server resolution would go in order of the name server list.  But it appears to be a little random (or I just dont understand the ordering).   Here is the Audit Trail for doing a bunch of lookups at the same time.  You can see it alternates between the network DNS server and Google DNS.   So just wanted to understand how your library chooses the lookup order.  Thanks.

;; Got answer:
;; ->>HEADER<<- opcode: Query, status: No Error, id: 20956

;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
01-08 12:51:40.554 I/mono-stdout(21142): ;; Got answer:;; QUESTION SECTION:
_minecraft._tcp.lobby.muttsworldmine.com. 	IN 	SRV

;; ANSWER SECTION:
_minecraft._tcp.lobby.muttsworldmine.com.299 	IN 	SRV 	0 5 25565 lobby.muttsworldmine.com.

01-08 12:51:40.554 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: No Error, id: 20956
;; Query time: 413 msec
;; SERVER: 8.8.4.4#53
;; WHEN: Mon Jan 08 17:51:38 Z 2018
;; MSG SIZE  rcvd: 113
01-08 12:51:40.554 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTORITY: 0, ADDITIONAL: 1
01-08 12:51:40.555 I/mono-stdout(21142): 
01-08 12:51:40.555 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:40.555 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:40.555 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:40.555 I/mono-stdout(21142): _minecraft._tcp.lobby.muttsworldmine.com. 	IN 	SRV
01-08 12:51:40.555 I/mono-stdout(21142): 
01-08 12:51:40.555 I/mono-stdout(21142): ;; ANSWER SECTION:
01-08 12:51:40.556 I/mono-stdout(21142): _minecraft._tcp.lobby.muttsworldmine.com.299 	IN 	SRV 	0 5 25565 lobby.muttsworldmine.com.
01-08 12:51:40.556 I/mono-stdout(21142): 
01-08 12:51:40.556 I/mono-stdout(21142): ;; Query time: 413 msec
01-08 12:51:40.556 I/mono-stdout(21142): ;; SERVER: 8.8.4.4#53
01-08 12:51:40.556 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:38 Z 2018
01-08 12:51:40.556 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 113
01-08 12:51:40.556 I/mono-stdout(21142): 
[0:] DNS Complete: lobby.muttsworldmine.com:158.69.97.57
01-08 12:51:40.699 I/Choreographer(21142): Skipped 40 frames!  The application may be doing too much work on its main thread.
01-08 12:51:40.730 D/ViewRootImpl@1201a22[SplashActivity](21142): mHardwareRenderer.destroy()#4
01-08 12:51:40.730 D/ViewRootImpl@1201a22[SplashActivity](21142): dispatchDetachedFromWindow
01-08 12:51:40.738 D/InputTransport(21142): Input channel destroyed: fd=69
[0:] Refresh: MCBF - Vanilla Server
192.168.1.1:5301-08 12:51:40.838 I/mono-stdout(21142): 192.168.1.1:53

[0:] Refresh: Official MineChat Server
192.168.1.1:5301-08 12:51:40.913 I/mono-stdout(21142): 192.168.1.1:53

01-08 12:51:40.957 I/art     (21142): Starting a blocking GC Explicit
01-08 12:51:41.013 I/art     (21142): Explicit concurrent mark sweep GC freed 5509(563KB) AllocSpace objects, 2(40KB) LOS objects, 54% free, 3MB/7MB, paused 791us total 55.452ms
01-08 12:51:41.014 D/Mono    (21142): GC_TAR_BRIDGE bridges 95 objects 1525 opaque 856 colors 95 colors-bridged 95 colors-visible 95 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.09ms tarjan 1.53ms scc-setup 0.09ms gather-xref 0.01ms xref-setup 0.00ms cleanup 0.10ms
01-08 12:51:41.014 D/Mono    (21142): GC_BRIDGE: Complete, was running for 58.07ms
01-08 12:51:41.014 D/Mono    (21142): GC_MINOR: (Nursery full) time 22.39ms, stw 23.54ms promoted 907K major size: 1872K in use: 1190K los size: 1024K in use: 331K
[0:] Refresh: MianiteLegacyMC
192.168.1.1:5301-08 12:51:41.090 I/mono-stdout(21142): 192.168.1.1:53

01-08 12:51:41.132 D/Mono    (21142): Assembly Ref addref Newtonsoft.Json[0xa09ef080] -> System.Diagnostics.Debug[0x86cee620]: 4
[0:] Refresh: Kongo PvP
192.168.1.1:5301-08 12:51:41.177 I/mono-stdout(21142): 192.168.1.1:53

[0:] Refresh: Novaria
192.168.1.1:5301-08 12:51:41.266 I/mono-stdout(21142): 192.168.1.1:53

01-08 12:51:41.268 D/ViewRootImpl@81a2e8e[MainActivity](21142): MSG_WINDOW_FOCUS_CHANGED 1
01-08 12:51:41.269 D/ViewRootImpl@81a2e8e[MainActivity](21142): mHardwareRenderer.initializeIfNeeded()#2 mSurface={isValid=true -2043650048}
01-08 12:51:41.274 V/InputMethodManager(21142): Starting input: tba=android.view.inputmethod.EditorInfo@d982c5f nm : com.kellyproductions.minechat ic=null
01-08 12:51:41.274 I/InputMethodManager(21142): startInputInner - mService.startInputOrWindowGainedFocus
01-08 12:51:41.276 D/InputTransport(21142): Input channel constructed: fd=84
01-08 12:51:41.276 D/InputTransport(21142): Input channel destroyed: fd=78
01-08 12:51:41.277 E/ViewRootImpl(21142): sendUserActionEvent() returned.
01-08 12:51:41.281 I/Choreographer(21142): Skipped 34 frames!  The application may be doing too much work on its main thread.
; (3 server found)
01-08 12:51:41.590 I/mono-stdout(21142): ; (3 server found);; Got answer:
;; ->>HEADER<<- opcode: Query, status: No Error, id: 20958
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
_minecraft._tcp.play.mcbf.pw.    	IN 	SRV

;; ANSWER SECTION:
_minecraft._tcp.play.mcbf.pw.   1800 	IN 	SRV 	0 1 25580 mc15.fadehost.com.

;; Query time: 183 msec
;; SERVER: 192.168.1.1#53
;; WHEN: Mon Jan 08 17:51:41 Z 2018
;; MSG SIZE  rcvd: 9401-08 12:51:41.591 I/mono-stdout(21142): ;; Got answer:

01-08 12:51:41.592 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: No Error, id: 20958
01-08 12:51:41.592 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTORITY: 0, ADDITIONAL: 1
01-08 12:51:41.592 I/mono-stdout(21142): 
01-08 12:51:41.592 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:41.592 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:41.592 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:41.593 I/mono-stdout(21142): _minecraft._tcp.play.mcbf.pw.    	IN 	SRV
01-08 12:51:41.593 I/mono-stdout(21142): 
01-08 12:51:41.593 I/mono-stdout(21142): ;; ANSWER SECTION:
01-08 12:51:41.593 I/mono-stdout(21142): _minecraft._tcp.play.mcbf.pw.   1800 	IN 	SRV 	0 1 25580 mc15.fadehost.com.
01-08 12:51:41.593 I/mono-stdout(21142): 
01-08 12:51:41.593 I/mono-stdout(21142): ;; Query time: 183 msec
01-08 12:51:41.593 I/mono-stdout(21142): ;; SERVER: 192.168.1.1#53
01-08 12:51:41.594 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.594 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 94
01-08 12:51:41.594 I/mono-stdout(21142): 
[0:] DNS Complete: play.mcbf.pw:167.114.4.25
; (3 server found)01-08 12:51:41.636 I/mono-stdout(21142): ; (3 server found)

;; Got answer:
;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20959
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1

;; ERROR: Non-Existent Domain
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
_minecraft._tcp.minechat.town.   	IN 	SRV

;; AUTHORITIES SECTION:
minechat.town.                  1799 	IN 	SOA 	a.ns.joker.com. hostmaster.joker.com. 2015082172 10240 7200 1209600 3600

;; Query time: 213 msec
;; SERVER: 8.8.4.4#53
;; WHEN: Mon Jan 08 17:51:41 Z 2018
;; MSG SIZE  rcvd: 119
01-08 12:51:41.636 I/mono-stdout(21142): ;; Got answer:
01-08 12:51:41.637 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20959
01-08 12:51:41.637 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1
01-08 12:51:41.637 I/mono-stdout(21142): 
[0:] DNS Complete: minechat.town:23.91.4.4601-08 12:51:41.637 I/mono-stdout(21142): ;; ERROR: Non-Existent Domain

01-08 12:51:41.637 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:41.637 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:41.637 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:41.637 I/mono-stdout(21142): _minecraft._tcp.minechat.town.   	IN 	SRV; (3 server found)
01-08 12:51:41.637 I/mono-stdout(21142): 

01-08 12:51:41.637 I/mono-stdout(21142): ;; AUTHORITIES SECTION:
;; Got answer:
;; ->>HEADER<<- opcode: Query, status: No Error, id: 20960
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
_minecraft._tcp.s48.hosthorde.com. 	IN 	SRV

;; AUTHORITIES SECTION:
hosthorde.com.                  1799 	IN 	SOA 	sid.ns.cloudflare.com. dns.cloudflare.com. 2026301719 10000 2400 604800 3600

;; Query time: 40 msec
01-08 12:51:41.637 I/mono-stdout(21142): minechat.town.                  1799 	IN 	SOA 	a.ns.joker.com. hostmaster.joker.com. 2015082172 10240 7200 1209600 3600
01-08 12:51:41.637 I/mono-stdout(21142): ;; SERVER: 8.8.4.4#53
;; WHEN: Mon Jan 08 17:51:41 Z 2018
;; MSG SIZE  rcvd: 120

01-08 12:51:41.637 I/mono-stdout(21142): ;; Query time: 213 msec
01-08 12:51:41.637 I/mono-stdout(21142): ;; SERVER: 8.8.4.4#53
01-08 12:51:41.637 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.637 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 119
01-08 12:51:41.638 I/mono-stdout(21142): 
01-08 12:51:41.648 I/mono-stdout(21142): ; (3 server found)
01-08 12:51:41.648 I/mono-stdout(21142): ;; Got answer:
01-08 12:51:41.648 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: No Error, id: 20960
01-08 12:51:41.648 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1
01-08 12:51:41.648 I/mono-stdout(21142): 
01-08 12:51:41.648 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:41.649 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:41.649 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:41.649 I/mono-stdout(21142): _minecraft._tcp.s48.hosthorde.com. 	IN 	SRV
01-08 12:51:41.649 I/mono-stdout(21142): 
01-08 12:51:41.649 I/mono-stdout(21142): ;; AUTHORITIES SECTION:
01-08 12:51:41.649 I/mono-stdout(21142): hosthorde.com.                  1799 	IN 	SOA 	sid.ns.cloudflare.com. dns.cloudflare.com. 2026301719 10000 2400 604800 3600
01-08 12:51:41.649 I/mono-stdout(21142): 
01-08 12:51:41.649 I/mono-stdout(21142): ;; Query time: 40 msec
01-08 12:51:41.649 I/mono-stdout(21142): ;; SERVER: 8.8.4.4#53
01-08 12:51:41.649 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.649 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 120
01-08 12:51:41.649 I/mono-stdout(21142): 
[0:] DNS Complete: s48.hosthorde.com:198.178.127.44
; (3 server found)
;; Got answer:
;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20961
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 101-08 12:51:41.701 I/mono-stdout(21142): ; (3 server found)

01-08 12:51:41.701 I/mono-stdout(21142): ;; Got answer:;; ERROR: Non-Existent Domain
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
_minecraft._tcp.kongopvp.com.    	IN 	SRV

;; AUTHORITIES SECTION:
kongopvp.com.                   600 	IN 	SOA 	ns35.domaincontrol.com. dns.jomax.net. 2017122204 28800 7200 604800 600

;; Query time: 57 msec
;; SERVER: 192.168.1.1#53
;; WHEN: Mon Jan 08 17:51:41 Z 2018
;; MSG SIZE  rcvd: 125

01-08 12:51:41.701 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20961
01-08 12:51:41.701 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1
01-08 12:51:41.702 I/mono-stdout(21142): 
01-08 12:51:41.702 I/mono-stdout(21142): ;; ERROR: Non-Existent Domain
01-08 12:51:41.702 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:41.702 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:41.702 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:41.702 I/mono-stdout(21142): _minecraft._tcp.kongopvp.com.    	IN 	SRV
01-08 12:51:41.702 I/mono-stdout(21142): 
01-08 12:51:41.702 I/mono-stdout(21142): ;; AUTHORITIES SECTION:
01-08 12:51:41.702 I/mono-stdout(21142): kongopvp.com.                   600 	IN 	SOA 	ns35.domaincontrol.com. dns.jomax.net. 2017122204 28800 7200 604800 600
01-08 12:51:41.702 I/mono-stdout(21142): 
01-08 12:51:41.702 I/mono-stdout(21142): ;; Query time: 57 msec
01-08 12:51:41.702 I/mono-stdout(21142): ;; SERVER: 192.168.1.1#53
01-08 12:51:41.702 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.702 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 125
01-08 12:51:41.702 I/mono-stdout(21142): 
[0:] DNS Complete: kongopvp.com:198.50.183.187
; (3 server found)01-08 12:51:41.765 I/mono-stdout(21142): ; (3 server found)

;; Got answer:
;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20962
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1

;; ERROR: Non-Existent Domain
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
_minecraft._tcp.play.novaria.world. 	IN 	SRV

;; AUTHORITIES SECTION:
novaria.world.                  1799 	IN 	SOA 	pdns1.registrar-servers.com. hostmaster.registrar-servers.com. 2017120602 43200 3600 604800 3601

;; Query time: 33 msec
;; SERVER: 8.8.4.4#53
;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.766 I/mono-stdout(21142): ;; Got answer:;; MSG SIZE  rcvd: 137

01-08 12:51:41.766 I/mono-stdout(21142): ;; ->>HEADER<<- opcode: Query, status: Non-Existent Domain, id: 20962
01-08 12:51:41.766 I/mono-stdout(21142): ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTORITY: 1, ADDITIONAL: 1
01-08 12:51:41.766 I/mono-stdout(21142): 
01-08 12:51:41.766 I/mono-stdout(21142): ;; ERROR: Non-Existent Domain
01-08 12:51:41.766 I/mono-stdout(21142): ;; OPT PSEUDOSECTION:
01-08 12:51:41.766 I/mono-stdout(21142): ; EDNS: version: 0, flags:; udp: 512
01-08 12:51:41.766 I/mono-stdout(21142): ;; QUESTION SECTION:
01-08 12:51:41.766 I/mono-stdout(21142): _minecraft._tcp.play.novaria.world. 	IN 	SRV
01-08 12:51:41.767 I/mono-stdout(21142): 
01-08 12:51:41.767 I/mono-stdout(21142): ;; AUTHORITIES SECTION:
01-08 12:51:41.767 I/mono-stdout(21142): novaria.world.                  1799 	IN 	SOA 	pdns1.registrar-servers.com. hostmaster.registrar-servers.com. 2017120602 43200 3600 604800 3601
01-08 12:51:41.767 I/mono-stdout(21142): 
01-08 12:51:41.767 I/mono-stdout(21142): ;; Query time: 33 msec
01-08 12:51:41.767 I/mono-stdout(21142): ;; SERVER: 8.8.4.4#53
01-08 12:51:41.767 I/mono-stdout(21142): ;; WHEN: Mon Jan 08 17:51:41 Z 2018
01-08 12:51:41.767 I/mono-stdout(21142): ;; MSG SIZE  rcvd: 137
01-08 12:51:41.767 I/mono-stdout(21142): 
[0:] DNS Complete: play.novaria.world:172.93.238.121

**Comment by gregmac** [comment]
Just ran into this myself. The way [LookupClient.GetNextServers()](https://github.com/MichaCo/DnsClient.NET/blob/63ca18adb333114fc09db1a9edb2d3f6dc3eceb7/src/DnsClient/LookupClient.cs#L652) works is it cycles through the servers list, and it retains this ordering for subsequent queries.

This means if you have name servers A,B,C,D, and then do 4 queries, they'll use the ordering:

1. B,C,D,A
2. C,D,A,B
3. D,A,B,C

Importantly, the first time you use it, it will actually start with the _second_ name server.

Cycling is clearly intended, but possibly not starting with the first server could be considered a bug IMHO -- it's certainly unexpected.

**Comment by MichaCo** [comment]
Good point, never thought about it though as it, in theory, doesn't really matter if it starts with the first or not. All servers will be used eventually anyways ^^.

Any yes, cycling though the configured servers is by design, for load balancing / reducing traffic on only one server if there are more than one...
The logic is pretty simple though.

I should probably `ToArray` at line https://github.com/MichaCo/DnsClient.NET/blob/63ca18adb333114fc09db1a9edb2d3f6dc3eceb7/src/DnsClient/LookupClient.cs#L657 so that the shuffle happens afterwards. Right now, you actually get the second server because of the lazy IEnumerable I think
Wasn't intentional

**Comment by MichaCo** [comment]
@vbisbest 
I just added some more functionality to `ILookupClient`:
I added the flag `PreserveNameServerOrder`. 
If enabled, it will always use the first server first and retry the next ones only in case of errors and such. Each consecutive call will also start the fallback with the first server. Only if the client decides to disable the first server, it then will always start with the second, and so on...
see #20 for more details

/closing this one for now
